### PR TITLE
[SPARK-45393][BUILD] Upgrade Hadoop to 3.4.0

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to upgrade to Apache Hadoop 3.4.0 for Apache Spark 4.0.0.

### Why are the changes needed?

To bring the new features like the following
- https://hadoop.apache.org/docs/r3.4.0
    - [HADOOP-18995](https://issues.apache.org/jira/browse/HADOOP-18995) Upgrade AWS SDK version to 2.21.33 for `S3 Express One Zone`
    - [HADOOP-18996](https://issues.apache.org/jira/browse/HADOOP-18996) S3A to provide full support for `S3 Express One Zone`
    - [HADOOP-18328](https://issues.apache.org/jira/browse/HADOOP-18328) Supports `S3 on Outposts`

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass the CIs.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @LuciferYang . I tried to search your previous PR , but couldn't find it. So, I recreated with your coauthor ship here.

If you want to re-open your PR, please let me know, @LuciferYang .

**Comment by dongjoon-hyun** [comment]
BTW, I'll add the following based on the failed module.
```
    <dependency>
      <groupId>org.bouncycastle</groupId>
      <artifactId>bcpkix-jdk18on</artifactId>
      <version>${bouncycastle.version}</version>
      <scope>test</scope>
    </dependency>
```

File: dev/deps/spark-deps-hadoop-3-hive-2.3 [hunk 1]
```diff
@@ -24,14 +24,14 @@ audience-annotations/0.12.0//audience-annotations-0.12.0.jar
...
 avro/1.11.3//avro-1.11.3.jar
-aws-java-sdk-bundle/1.12.367//aws-java-sdk-bundle-1.12.367.jar
 azure-data-lake-store-sdk/2.3.9//azure-data-lake-store-sdk-2.3.9.jar
 azure-keyvault-core/1.0.0//azure-keyvault-core-1.0.0.jar
 azure-storage/7.0.1//azure-storage-7.0.1.jar
 blas/3.0.3//blas-3.0.3.jar
 bonecp/0.8.0.RELEASE//bonecp-0.8.0.RELEASE.jar
 breeze-macros_2.13/2.1.0//breeze-macros_2.13-2.1.0.jar
 breeze_2.13/2.1.0//breeze_2.13-2.1.0.jar
+bundle/2.23.19//bundle-2.23.19.jar
```
Comment (dongjoon-hyun) [inline dev/deps/spark-deps-hadoop-3-hive-2.3:34]: https://mvnrepository.com/artifact/software.amazon.awssdk/bundle/2.23.19

**Comment by LuciferYang** [comment]
> Hi, @LuciferYang . I tried to search your previous PR , but couldn't find it. So, I recreated with your coauthor ship here.
> 
> If you want to re-open your PR, please let me know, @LuciferYang .

Thanks @dongjoon-hyun ~ Let's use this one ~

**Comment by dongjoon-hyun** [comment]
Got it. Thank you, @LuciferYang .

**Review by LuciferYang — APPROVED** [review]
+1, LGTM (pending test)

**Comment by dongjoon-hyun** [comment]
Thank you, @LuciferYang .

For this PR, I fixed the following three so far, but I guess there is one more to go. Let's see the CI result.
- #45585
- #45594
- #45597

File: resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala [hunk 2]
```diff
@@ -292,7 +292,8 @@ class YarnClusterSuite extends BaseYarnClusterSuite {
     }
   }
 
-  test("running Spark in yarn-cluster mode displays driver log links") {
+  // TODO(SPARK-47491): Re-enable `driver log links` test in YarnClusterSuite
+  ignore("running Spark in yarn-cluster mode displays driver log links") {
```
Comment (dongjoon-hyun) [inline resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala:296]: I'll handle this later in SPARK-47491.

**Comment by dongjoon-hyun** [comment]
At the previous commit (https://github.com/dongjoon-hyun/spark/runs/22865082407), we already passed all tests except one YARN failure. And, the failed test case is ignored here. Let me merge this to move forward.

Thank you, @LuciferYang and @yaooqinn .

File: dev/deps/spark-deps-hadoop-3-hive-2.3 (hunk 1 above)
Comment (yaooqinn) [reply dev/deps/spark-deps-hadoop-3-hive-2.3:34]: image
What a big jar...:)

Comment (LuciferYang) [reply dev/deps/spark-deps-hadoop-3-hive-2.3:34]: .... surprising

**Comment by LuciferYang** [comment]
Sorry to disturb everyone, but when I execute `OrcEncryptionSuite` on my M2 Max, I find that there are some differences when using Hadoop 3.4.0 and Hadoop 3.3.4.

`build/sbt clean "sql/testOnly org.apache.spark.sql.execution.datasources.orc.OrcEncryptionSuite"`

- branch-3.5(with hadoop 3.3.4)

```
[info] OrcEncryptionSuite:
14:44:11.580 WARN org.apache.hadoop.util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
[info] - Write and read an encrypted file (1 second, 921 milliseconds)
[info] - Write and read an encrypted table (374 milliseconds)
[info] - SPARK-35325: Write and read encrypted nested columns (358 milliseconds)
[info] - SPARK-35992: Write and read fully-encrypted columns with default masking (570 milliseconds)
14:44:15.461 WARN org.apache.spark.sql.execution.datasources.orc.OrcEncryptionSuite: 

[info] Run completed in 4 seconds, 694 milliseconds.
[info] Total number of tests run: 4
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 4, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

- master(with hadoop 3.4.0)

```
[info] OrcEncryptionSuite:
14:49:15.267 WARN org.apache.hadoop.util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
14:49:17.636 WARN org.apache.hadoop.crypto.OpensslCipher: Failed to load OpenSSL Cipher.
java.lang.UnsatisfiedLinkError: 'boolean org.apache.hadoop.util.NativeCodeLoader.buildSupportsOpenssl()'
	at org.apache.hadoop.util.NativeCodeLoader.buildSupportsOpenssl(Native Method)
	at org.apache.hadoop.crypto.OpensslCipher.<clinit>(OpensslCipher.java:86)
	at org.apache.hadoop.crypto.OpensslAesCtrCryptoCodec.<init>(OpensslAesCtrCryptoCodec.java:36)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
[info] - Write and read an encrypted file (2 seconds, 343 milliseconds)
[info] - Write and read an encrypted table (405 milliseconds)
[info] - SPARK-35325: Write and read encrypted nested columns (308 milliseconds)
[info] - SPARK-35992: Write and read fully-encrypted columns with default masking (555 milliseconds)
14:49:19.493 WARN org.apache.spark.sql.execution.datasources.orc.OrcEncryptionSuite: 

[info] Run completed in 5 seconds, 84 milliseconds.
[info] Total number of tests run: 4
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 4, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

When using Hadoop 3.4.0, although there were no test failures, an `UnsatisfiedLinkError` was thrown. Is this expected, or do I need to configure additional dependencies? This issue should only occur on Apple Silicon chips now. @dongjoon-hyun @steveloughran

**Comment by dongjoon-hyun** [comment]
To @LuciferYang , have you try other combination like `branch-3.4 + Hadoop 3.4` or `master + 3.3.4`? It seems that you are reporting only the difference on branches which could be a result of many other stuffs like ORC version difference or installed ssl versions.

>  I find that there are some differences when using Hadoop 3.4.0 and Hadoop 3.3.4.

In short, I'm wondering if you are reporting that the result before this commit and after this commit, @LuciferYang . For your claim, could you share us that result?

**Comment by LuciferYang** [comment]
@dongjoon-hyun Test on master:

before this commit:

```
git reset --hard a34c8ceb19bd1c1548a60bb144d1c587a2861cd8 // [SPARK-47462][SQL] Align mappings of other unsigned numeric types with TINYINT in MySQLDialect
build/sbt clean "sql/testOnly org.apache.spark.sql.execution.datasources.orc.OrcEncryptionSuite"
```

result:
```
[info] OrcEncryptionSuite:
12:47:33.148 WARN org.apache.hadoop.util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
[info] - Write and read an encrypted file (2 seconds, 117 milliseconds)
[info] - Write and read an encrypted table (375 milliseconds)
[info] - SPARK-35325: Write and read encrypted nested columns (275 milliseconds)
[info] - SPARK-35992: Write and read fully-encrypted columns with default masking (517 milliseconds)
12:47:37.058 WARN org.apache.spark.sql.execution.datasources.orc.OrcEncryptionSuite: 

===== POSSIBLE THREAD LEAK IN SUITE o.a.s.sql.execution.datasources.orc.OrcEncryptionSuite, threads: ForkJoinPool.commonPool-worker-4 (daemon=true), rpc-boss-3-1 (daemon=true), Thread-17 (daemon=true), ForkJoinPool.commonPool-worker-2 (daemon=true), shuffle-boss-6-1 (daemon=true), ForkJoinPool.commonPool-worker-1 (daemon=true), Thread-18 (daemon=true), ForkJoinPool.commonPool-worker-3 (daemon=true) =====
[info] Run completed in 4 seconds, 760 milliseconds.
[info] Total number of tests run: 4
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 4, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```


after this commit:

```
git reset --hard 49b4c3bc9c09325de941dfaf41e4fd3a4a4c345f // [SPARK-45393][BUILD] Upgrade Hadoop to 3.4.0
build/sbt clean "sql/testOnly org.apache.spark.sql.execution.datasources.orc.OrcEncryptionSuite"
```

result:

```
[info] OrcEncryptionSuite:
12:42:55.441 WARN org.apache.hadoop.util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
12:42:57.950 WARN org.apache.hadoop.crypto.OpensslCipher: Failed to load OpenSSL Cipher.
java.lang.UnsatisfiedLinkError: 'boolean org.apache.hadoop.util.NativeCodeLoader.buildSupportsOpenssl()'
	at org.apache.hadoop.util.NativeCodeLoader.buildSupportsOpenssl(Native Method)
	at org.apache.hadoop.crypto.OpensslCipher.<clinit>(OpensslCipher.java:86)
	at org.apache.hadoop.crypto.OpensslAesCtrCryptoCodec.<init>(OpensslAesCtrCryptoCodec.java:36)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
[info] - Write and read an encrypted file (2 seconds, 486 milliseconds)
[info] - Write and read an encrypted table (402 milliseconds)
[info] - SPARK-35325: Write and read encrypted nested columns (299 milliseconds)
[info] - SPARK-35992: Write and read fully-encrypted columns with default masking (623 milliseconds)
12:42:59.856 WARN org.apache.spark.sql.execution.datasources.orc.OrcEncryptionSuite: 

===== POSSIBLE THREAD LEAK IN SUITE o.a.s.sql.execution.datasources.orc.OrcEncryptionSuite, threads: rpc-boss-3-1 (daemon=true), Thread-17 (daemon=true), ForkJoinPool.commonPool-worker-2 (daemon=true), shuffle-boss-6-1 (daemon=true), ForkJoinPool.commonPool-worker-1 (daemon=true), Thread-18 (daemon=true), ForkJoinPool.commonPool-worker-3 (daemon=true) =====
[info] Run completed in 5 seconds, 291 milliseconds.
[info] Total number of tests run: 4
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 4, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

**Comment by dongjoon-hyun** [comment]
Ack. Thank you for sharing. Let me take a look at that as a independent JIRA issue because it's Mac only issue, @LuciferYang .

**Comment by dongjoon-hyun** [comment]
[SPARK-49055](https://issues.apache.org/jira/browse/SPARK-49055) is filed, @LuciferYang .

**Comment by LuciferYang** [comment]
> Ack. Thank you for sharing. Let me take a look at that as a independent JIRA issue because it's Mac only issue, @LuciferYang .

@dongjoon-hyun I apologize for providing misleading information. I just reviewed the recent GA test logs and I found that this is not a Mac Only issue:

- https://github.com/apache/spark/actions/runs/10155611310/job/28082653276

image

**Comment by dongjoon-hyun** [comment]
To @LuciferYang , according to the Hadoop code, HADOOP-17982 seems to change the log level only at Hadoop 3.4.0.

https://github.com/apache/hadoop/pull/3599/files

**Comment by dongjoon-hyun** [comment]
There is no functional change in Hadoop code since last 10 year except the above log level change. Given that, we can ignore the warning message. WDYT, @LuciferYang ?

**Comment by LuciferYang** [comment]
@dongjoon-hyun Thank you for your explanation, agree with you

**Pull request closed without merging**